### PR TITLE
[WIP] Improve deployment strategy

### PR DESCRIPTION
**Pull request by davidgasquez**
Hey there! Sending this early WIP PR so we can have a place to discuss the changes and status of these changes.

The goal is to send raw deserialized protobufs as JSON records to BigQuery using the Python client. Since we'll be working on production data this step must catch all the exceptions and avoid affecting Kiner.

### CODE REVIEW

**Comment by davidgasquez** [comment]
A few changes worth mentioning that are already in this PR:

- No more `.env`. The Makefile will reuse the user local credentials 
https://github.com/bufferapp/buda-events-collector/blob/d5cc6414cc2df1c5ef1e975b3d7c7b4268ce5c7f/Makefile#L2
- [Up to date base image](https://github.com/bufferapp/buda-events-collector/pull/20/commits/65e570d6ffb255a5635517f068f660ef4b008004). The image we were using is not really maintained. Also, is now possible to install `grpc` more easily so we can use a standard Python base image.
- Cleaned up the autoreload file. I wasn't sure we were using it. If we are I'd love to explore better alternatives to avoid an external script. 
- Some :art:! 
- Fixed the requirements and upgraded `grpc` from `1.4` to 1.19`
- Removed signal handling. This is done by Kiner and grpc so we don't need to do that ourselves.

**Comment by davidgasquez** [comment]
I'd love to pick your mind in two topics @michael-erasmus! :smile: 

1. Does it makes sense to send all the records to the same table in `bq`. This could be something like `buda-events` that we can partition by date and cluster by event type.
1. Can we simplify or remove the client side? I think the client is creating a bunch of events from the same package and sending them over. This makes it hard to maintain since if the definitions change we'll forget to change them here. Also, we're using the same Python package so most of this should be already tested by `grpc`. I think we could simplify that a lot and have only a `test_server.py` file that sends a few tests events to check we're not crashing at any point. We can add a switch to avoid sending the test data to Kinesis and BQ. 

I also had an extra question but I think we can leave it for later. I was thinking that perhaps we can change the collector a bit. Right now we have an `rpc` for each event type and we're duplicating a bunch of code. Not sure if it'll be possible or easy but I was thinking to move more towards a `CollectEvent` generic RPC. We can define the events we track in the `gRPC` definitions inside an ENUM or similar.

Would love to hear how all of this sounds to you!

**Comment by michael-erasmus** [comment]
Hi @davidgasquez!

Looking good so far!

Great questions, let's jump into that:

>  Does it makes sense to send all the records to the same table in bq. This could be something like buda-events that we can partition by date and cluster by event type.

Yes that could work I think (I couldn't help but chuckle a bit at how similar this approach seems to our legacy `event_data` setup, obviously without the Redshift limitations). So the idea then would be that we store a json blob that we can extract in subsequent incremental dbt models,

> Can we simplify or remove the client side?  

> We can add a switch to avoid sending the test data to Kinesis and BQ.

I've mainly used the test client in the past as a way to test new event pipelines end-to-end before deploying the events collector, as a way to also test the lambda clients and firehoses. 

I'm not sure if it's worth keeping it around for now, or maybe it could even be a separate repo? There are probably better ways to handle testing and this has been a very manual/hacky approach so I'm totally open to maybe cleaning things up and handling testing in a better way, so I totally trust your judgement on this!

File: client/client/actions_taken.py [hunk 1]
```diff
@@ -5,14 +5,14 @@
...
+        application="my-awesome-app",
+        location="awesome_page",
+        action="amazing_button_clicked",
     )
 
     event.created_at.GetCurrentTime()
-    event.metadata['userId'] = new_uuid().id
-    event.metadata['foo'] = 'bar'
+    event.metadata["userId"] = new_uuid().id
+    event.metadata["foo"] = "bar"
```
[the remark this answers is not in the document]
Comment (davidgasquez) [reply client/client/actions_taken.py]: Yes! :sweat_smile: Apologies for the noise it adds to changes in old repositories with the first change. I think is worth it though!

**Comment by davidgasquez** [comment]
> Yes that could work I think (I couldn't help but chuckle a bit at how similar this approach seems to our legacy event_data setup, obviously without the Redshift limitations). So the idea then would be that we store a json blob that we can extract in subsequent incremental dbt models, is that right?

Right! That, at least would be my preferred approach since it allows us to move faster. That said, since we'll be using Protobufs, we could also go the other path and have different tables with defined schemas. 

The first approach will let us manage the schemas at the `dbt` level, add some flexibility and speed (we won't need to change anything before adding new fields in the protobufs), but we'd be wasting some time  and space parsing and storing the JSONs blobs. The multiple table approach might be slower to start and could be also a bit harder to maintain the schemas over time but it'll optimize the space and it'll look cleaner. :sweat_smile: 

Does this sound good to you? Since I'll be focusing first on iterating the deployment strategy we can continue to discuss this trade off.

> I've mainly used the test client in the past as a way to test new event pipelines end-to-end before deploying the events collector, as a way to also test the lambda clients and firehoses.

That makes total sense and I wonder what would be the best approach to this issue. I'd love at least to automate the testing in some way and reduce the friction to add new tests. I'll give some thoughts to this in the following days!

**Comment by davidgasquez** [comment]
Quick update! The last commit added a health check to the service. This will enable smarter deployments! 

Besides that, I've also started to use the submodule approach since I was modifying the protobufs while testing the server. So far is been easier and faster to work with than the package approach. Ideally, we could do a multi-stage Dockerfile to build the Python package but right now everything is on the same one. 

I'm also aware this PR is quite huge. Perhaps we could merge it once the deployment strategy is figured out and apply it knowing that nothing significantly changed in the server. Then we can start working on sending data to BQ in the next PR.

**Comment by michael-erasmus** [comment]
> Perhaps we could merge it once the deployment strategy is figured out and apply it knowing that nothing significantly changed in the server. Then we can start working on sending data to BQ in the next PR.

That sounds good to me, having a better deployment strategy in place could provide value immediately and it could be great to break the PR's up into smaller chunks :+1:

**Comment by davidgasquez** [comment]
I think we should be all set after the latest commit. The final deployment has a few new things:
- `minReadySeconds` to wait 5 seconds until sending data to the pod
- `RollingUpdate` strategy. It'll create pods one by one only continuing if all of them are available. The first pod will be created without deleting any pod (`maxSurge`)
- `readinessProbe` and `livenessProbe` to handle the rolling update
-  Limit `resources`. Got the current usage using `kubectl top pods` and added a 25% more resources. I'm not sure if we should keep this or just leave everything unlimited.

I'd love to merge and deploy this. That said, I'll wait until monday morning to avoid any potential issues over the weekend.

**Comment by davidgasquez** [comment]
I was reviewing the changes and there is one thing that's bugging me. In the original deployment we aren't specifying any replicas and we see three pods running. I think we should perhaps make that explicit with `replicas: 3` just in case the defaults have changed.

**Comment by davidgasquez** [comment]
Just made the change. Everything looks good for me now! :+1:

**Comment by davidgasquez** [comment]
Merging and deploying! :rocket:

**Pull request merged**